src/mon/PGMonitor.cc

```diff
@@ -1234,6 +1235,146 @@ bool PGMonitor::preprocess_command(MonOpRequestRef op)
+
+      // get pgids to process and prune duplicates
+      cmd_getval(g_ceph_context, cmdmap, "pgid", pgidstr);
+      for (auto& pstr : pgidstr) {
+	pgidstr_nodup.insert(pstr);
+      }
+      if (pgidstr.size() != pgidstr_nodup.size()) {
+	// move elements only when there were duplicates, as this
+	// reorders them
+	pgidstr.resize(pgidstr_nodup.size());
+	auto it = pgidstr_nodup.begin();
+	for (size_t i = 0 ; i < pgidstr_nodup.size(); i++) {
+	  pgidstr[i] = std::move(*it++);
+	}
+      }
+
+      for (auto& pstr : pgidstr) {
+	pg_t parsed_pg;
+	if (!parsed_pg.parse(pstr.c_str())) {
+	  ss << "invalid pgid '" << pstr << "'; ";
+	  r = -EINVAL;
+	} else 	if (!pg_map.pg_stat.count(parsed_pg)) {
+	  ss << "pg " << pstr << " not exists; ";
+	  r = -ENOENT;
+	} else {
+	  pg_stat_t& workpg = pg_map.pg_stat[parsed_pg];
+
+	  // discard pgs for which user requests are pointless
+	  switch (actual_op)
+	  {
+	    case OFR_RECOVERY:
+	      if ((workpg.state & (PG_STATE_DEGRADED | PG_STATE_RECOVERY_WAIT | PG_STATE_RECOVERING)) == 0) {
+		ss << "pg " << pstr << " doesn't require recovery; ";
+		r = -EINVAL;
+		continue;
+	      } else  if (workpg.state & PG_STATE_FORCED_RECOVERY) {
+		ss << "pg " << pstr << " recovery already forced; ";
+		r = -EINVAL;
+		continue;
+	      }
+	      break;
+	    case OFR_BACKFILL:
+	      if ((workpg.state & (PG_STATE_DEGRADED | PG_STATE_BACKFILL_WAIT | PG_STATE_BACKFILL)) == 0) {
+		ss << "pg " << pstr << " doesn't require backfilling; ";
+		r = -EINVAL;
+		continue;
+	      } else  if (workpg.state & PG_STATE_FORCED_BACKFILL) {
+		ss << "pg " << pstr << " backfill already forced; ";
+		r = -EINVAL;
+		continue;
+	      }
+	      break;
+	    case OFR_BACKFILL | OFR_CANCEL:
+	      if ((workpg.state & PG_STATE_FORCED_BACKFILL) == 0) {
+		ss << "pg " << pstr << " backfill not forced; ";
+		r = -EINVAL;
+		continue;
+	      }
+	      break;
+	    case OFR_RECOVERY | OFR_CANCEL:
+	      if ((workpg.state & PG_STATE_FORCED_RECOVERY) == 0) {
+	        ss << "pg " << pstr << " recovery not forced; ";
+		r = -EINVAL;
+		continue;
+	      }
+	      break;
+	    default:
+	      assert(0 == "actual_op value is not supported");
+	  }
+
+	  parsed_pgs.push_back(std::move(parsed_pg));
+	}
+      }
+    }
+
+    // respond with error only when no pgs are correct
+    if (parsed_pgs.size() == 0) {
+      // yes, in case of mixed errors, only the last one will be emitted,
+      // but the message presented will be fine
+      goto reply;
+    } else {
+      // clear error to not confuse users/scripts
+      r = 0;
+    }
+
+    // optimize the command -> messages conversion, use only one message per distinct OSD
+    map<int, list<pg_t> > osdpgs;
+
+    // group pgs to process by osd
+    for (auto& pgid : parsed_pgs) {
+      pg_stat_t& workpg = pg_map.pg_stat[pgid];
+      set<int32_t> osds;
+      for (auto i : workpg.up) {
+	osds.insert(i);
+      }
+      for (auto i : workpg.acting) {
+	osds.insert(i);
+      }
```

Comment:
the two loops above can be shortened to:

    set<int32_t> osds(workpg.up.begin(), workpg.up.end());
    osds.insert(workpg.acting.begin(), workpg.acting.end());